File: exercises.java
Oppgave 1:

// get the country code
string countryCode = getCountryCode('REMOTE_ADDR');
 
// if country code is US
if (countryCode == 'US') {
    // display the form input for state
    println(formInputState());
}

---------------------------
---------------------------

Oppgave 2:

function canBuyBeer(age, money) {
  if(age >= 21 && money >= 20) {
    return true
  }
  return false
}

--------------------------
--------------------------

Oppgave 3:

function shouldShowImage(itemIndex, article, showAllImages) {
  return [0, 1, 2].includes(itemIndex)
    ? !!article.imageUrl
    : showAllImages
      ? !!article.imageUrl
      : false
}

--------------------------
--------------------------

Oppgave 4:


private static List readLines(String fileName) {
  String line;
  List file = new List();

  try {    
    BufferedReader in = new BufferedReader(new FileReader(fileName));
    while ((line = in.readLine()) != null)
      file.add(line);
    in.close();
  } catch (Exception e){
    System.out.println(e);
    return null;
  }
  return file;
}

---------------------------
---------------------------

Oppgave 5:


boolean userEmail = user.email.contains('computas');
if(userEmail) {
    addUser(user);
    alertUser();     
} else {
    alertUser();
}

---------------------------
---------------------------

Oppgave 6:

String name = "Ryan McDermott";

function splitIntoFirstAndLastName() {
  name = name.split(" ");
}
splitIntoFirstAndLastName();

---------------------------
---------------------------

Oppgave 7:

function getProduct(htmlResponse) {
   String title = htmlResponse.css('.title').html();
   String priceHTML = htmlResponse.css('.price').html();
   int price;

   try {
      price = int(priceHTML);
   } except ValueError {
      price = null;
   }

   String sizesHTML = list(htmlResponse.css('.product .size').html());
   String[] sizes = [parseSize(htmlSize) for htmlSize in sizesHTML];
   return sizes;
}

---------------------------
---------------------------

Oppgave 8:

Her har vi en enkel og noe uferdig versjon av MineSweeper. 

Her kan man anta at myTruth og myShow allerede er initialisert med verdier.
- MyTruth vil inneholde 0 i de tilfellene det ikke er noen mine på gitt index. 
- MyShow vil si noe om cellen på gitt index er skjult eller ikke.

public class MineSweeper {
  private int[][] myTruth;
	private boolean[][] myShow;
	
	public void cellPicked(int row, int col) {
    if (inBounds(row, col) && !myShow[row][col]) {
      myShow[row][col] = true;
		
			if (myTruth[row][col] == 0) {
        for (int r = -1; r <= 1; r++) {
					for (int c = -1; c <= 1; c++) {
						cellPicked(row + r, col + c);
          }
        }
			}	
		}
	}
	
	public boolean inBounds(int row, int col) {
    return 0 <= row && row < myTruth.length && 0 <= col && col < myTruth[0].length;
	}
}

---------------------------
---------------------------

Oppgave 9:

Her har vi en klasse som styrer hvilken mengde discount man kan få basert på gitte typer.
Vi har her veldig lite informasjon, så prøv så godt som du kan.

- calculate: Skal regne ut pricen for en gitt kunde.
- type: Skal her stå for en status om kunden som kan si hvor verdifull kunden er og dermed bestemmer hvilken discount de får. Vi vet veldig lite her, så vi kan bare anta 4 forskjellige nivåer
- years: Hvor mange år personen har vært kunde. Kunden får 1% men maksimalt 5 procent i loyalty discount.
- amount: Den opprinnelige prisen
- result: Dette er så den totale prisen med loyaltydiscount og kundediscount trukket fra

public class Class1 {
  public decimal calculate(decimal amount, int type, int years) {
    decimal result = 0;
    decimal disc = (years > 5) ? (decimal) 5 / 100 : (decimal) years / 100; 

    if (type == 1) {
      result = amount;
    } else if (type == 2) {
      result = (amount - (0.1m * amount)) - disc;
    } else if (type == 3) {
      result = (amount - (0.5m * amount)) - disc;
    } else if (type == 4) {
      result = (amount - (0.7m * amount) - disc;
    }
    return result;
  }
}

File: exercises.ts
// Oppgavene under (skrevet i typescript) bryter med en eller flere Clean Code prinsipper.
// Se hvor mange du klarer å identifisere. Prøv gjerne også å se om
// du klarer refaktorere dem.

// ---------------------------
// ---------------------------


// Oppgave 1:
function a() {
  // get the country code
  const countryCode: string = getCountryCode('REMOTE_ADDR');

  // if country code is US
  if (countryCode == 'US') {
    // display the form input for state
    console.log(formInputState());
  }
}

// ---------------------------
// ---------------------------

// Oppgave 2:

function canBuyBeer(age: number, money: number) {
  if (age >= 21 && money >= 20) {
    return true
  }
  return false
}

// --------------------------
// --------------------------

// Oppgave 3:

// Her er det brukt javascript ternaries som basically er det samme som if else.
// if = utrykk  
// ? = utrykk som evalures hvis true
// : = utrykk som evalueres hvis false

function shouldShowImage(itemIndex: number, article: {imageUrl: string}, showAllImages: boolean) {
  return [0, 1, 2].includes(itemIndex)
    ? article.imageUrl != null
    : showAllImages
      ? article.imageUrl != null
      : false
}

// --------------------------
// --------------------------

interface IUser {
  email: string
};

// Oppgave 4:
function f(user: IUser) {
  const userEmail: boolean = user.email.includes('computas');
  if (userEmail) {
    addUser(user);
    alertUser();
  } else {
    alertUser();
  }
}

// ---------------------------
// ---------------------------

// Oppgave 5:

let n: any = "Ryan McDermott";

function splitIntoFirstAndLastName() {
  n = n.split(" ");
}
splitIntoFirstAndLastName();

// ---------------------------
// ---------------------------

// Oppgave 6:

function getProduct(htmlResponse) {
  const title: string = htmlResponse.css('.title').html();
  const priceHTML: string = htmlResponse.css('.price').html();
  let price: Number;

  try {
    price = parseInt(priceHTML);
  } catch (e) {
    price = null;
  }

  const sizesHTML: string[] = [htmlResponse.css('.product .size').html()];
  const sizes: string[] = sizesHTML.map((htmlSize) => parseSize(htmlSize));

  return {
    sizes,
    price,
    title
  };
}

// ---------------------------
// ---------------------------

// ---------------------------
// Ekstraoppgaver
// ---------------------------

// Oppgave 7:

// Her har vi en enkel og noe uferdig versjon av MineSweeper. 

// Her kan man anta at myTruth og myShow allerede er initialisert med verdier.
// - MyTruth vil inneholde 0 i de tilfellene det ikke er noen mine på gitt index. 
// - MyShow vil si noe om cellen på gitt index er skjult eller ikke.



class MineSweeper {
  myTruth: Number[][];
  myShow: boolean[][];
  cellPicked(row: number, col: number): void {
    if (this.inBounds(row, col) && !this.myShow[row][col]) {
      this.myShow[row][col] = true;

      if (this.myTruth[row][col] == 0) {
        for (let r = -1; r <= 1; r++) {
          for (let c = -1; c <= 1; c++) {
            cellPicked(row + r, col + c);
          }
        }
      }
    }
  }

  inBounds(row: number, col: number): boolean {
    return 0 <= row && row < this.myTruth.length && 0 <= col && col < this.myTruth[0].length;
  }
}

// ---------------------------
// ---------------------------

// Oppgave 8:

// Her har vi en klasse som styrer hvilken mengde discount man kan få basert på gitte typer.
// Vi har her veldig lite informasjon, så prøv så godt som du kan.

// - calculate: Skal regne ut pricen for en gitt kunde.
// - type: Skal her stå for en status om kunden som kan si hvor verdifull kunden er og dermed bestemmer hvilken discount de får. Vi vet veldig lite her, så vi kan bare anta 4 forskjellige nivåer
// - years: Hvor mange år personen har vært kunde. Kunden får 1% men maksimalt 5 procent i loyalty discount.
// - amount: Den opprinnelige prisen
// - result: Dette er så den totale prisen med loyaltydiscount og kundediscount trukket fra

function calculate(amount: number, type: number, years: number): number {
  let result = 0;
  const disc = (years > 5) ? 5 / 100 : years / 100;

  if (type == 1) {
    result = amount;
  } else if (type == 2) {
    result = (amount - (0.1 * amount)) - disc;
  } else if (type == 3) {
    result = (amount - (0.5 * amount)) - disc;
  } else if (type == 4) {
    result = (amount - (0.7 * amount)) - disc;
  }
  return result;
}
